main/modes/mode_nvs_manager.c

```diff
@@ -0,0 +1,3608 @@
+    resetMeleeMenu(nvsManager->menu, modeNvsManager.modeName, nvsManagerTopLevelCb);
+    addRowToMeleeMenu(nvsManager->menu, str_summary);
+    addRowToMeleeMenu(nvsManager->menu, str_manage_data);
+    addRowToMeleeMenu(nvsManager->menu, str_send_all);
+    addRowToMeleeMenu(nvsManager->menu, str_receive);
+
+    // Add the row for factory resetting the Swadge
+    if (nvsManager->eraseDataSelected)
+    {
+        if (nvsManager->eraseDataConfirm)
+        {
+            addRowToMeleeMenu(nvsManager->menu, str_confirm_yes);
+        }
+        else
+        {
+            addRowToMeleeMenu(nvsManager->menu, str_confirm_no);
+        }
+    }
+    else
+    {
+        addRowToMeleeMenu(nvsManager->menu, str_factory_reset);
+    }
+
+    addRowToMeleeMenu(nvsManager->menu, str_exit);
+
+    // Set the position
+    if(resetPos)
+    {
+        nvsManager->topLevelPos = 0;
+    }
+    nvsManager->menu->selectedRow = nvsManager->topLevelPos;
+    nvsManager->menu->usePerRowXOffsets = true;
+
+    nvsManager->screen = NVS_MENU;
+}
+
+/**
+ * Callback for the top level menu
+ *
+ * @param opt The menu option which was selected
+ */
+void nvsManagerTopLevelCb(const char* opt)
+{
+    // Save the position
+    nvsManager->topLevelPos = nvsManager->menu->selectedRow;
+
+    // Handle the option
+    if(str_summary == opt)
+    {
+        nvsManager->screen = NVS_SUMMARY;
+    }
+    else if(str_manage_data == opt)
+    {
+        if(!nvsManager->warningAccepted)
+        {
+            nvsManager->fnNextMenu = nvsManagerSetUpManageDataMenu;
+            nvsManager->screen = NVS_WARNING;
+        }
+        else
+        {
+            nvsManagerSetUpManageDataMenu(true);
+        }
+    }
+    else if(str_send_all == opt)
+    {
+        nvsManager->sendMode = NVS_SEND_MODE_ALL;
+        nvsManager->fnReturnMenu = nvsManagerSetUpTopMenu;
+        nvsManagerSetUpSendConnMenu(true);
+    }
+    else if(str_receive == opt)
+    {
+        if(!nvsManager->warningAccepted)
+        {
+            nvsManager->fnNextMenu = nvsManagerSetUpReceiveMenu;
+            nvsManager->screen = NVS_WARNING;
+        }
+        else
+        {
+            nvsManagerSetUpReceiveMenu(true);
+        }
+    }
+    else if(str_factory_reset == opt)
+    {
+        nvsManager->eraseDataSelected = true;
+        nvsManagerSetUpTopMenu(false);
+    }
+    else if (str_confirm_yes == opt)
+    {
+        // If this succeeds, we shouldn't let someone back into the main menu, so switch to test mode immediately
+        // If this fails, we have no idea what state NVS is in, so send them back to the main menu and pray
+        // The emulator will encounter issues too unless it's immediately closed
+        if(eraseNvs())
+        {
+#ifdef EMU
+            emu_quit();
+#else
+            switchToSwadgeMode(&modeTest);
+#endif
+        }
+        else
+        {
+#ifdef EMU
+            exit(1);
+#else
+            switchToSwadgeMode(&modeMainMenu);
+#endif
+        }
```

Comment:
I needed to include `emu_main.h` in order to use `emu_quit()` here, since switching modes doesn't reset anything on the emu, and I believe there became an issue with the json file after erasing it unless the emu was restarted

Reply:
The emu doesn't keep any file handles open, and it doesn't cache info from NVM the way we do on hardware, so I'm not sure what your issue was or what you're trying to reset. It still feels weird to me, and in general the fewer exceptions we make for the emu, the better.

Reply:
Tried again and can't reproduce anymore. Removed the code and import.